CONTRIBUTING.md

```diff
@@ -9,70 +36,187 @@ license the code under the same terms as the project's overall 'outbound'
+Synapse. [docs/dev](docs/dev) exists primarily to house documentation for
+Synapse developers. [docs/admin_api](docs/admin_api) houses documentation
+regarding Synapse's Admin API, which is used mostly by sysadmins and external
+service developers.
+
+If you add new files added to either of these folders, please use [Github-Flavoured
+Markdown](https://guides.github.com/features/mastering-markdown/).
 
-```
-# Install the dependencies
-pip install -e ".[lint,mypy]"
+Some documentation also exists in [Synapse's Github
+Wiki](https://github.com/matrix-org/synapse/wiki), although this is primarily
+contributed to by community authors.
+
+
+# 7. Test, test, test!
+<a name="test-test-test"></a>
+
+While you're developing and before submitting a patch, you'll
+want to test your code.
 
-# Run the linter script
+## Run the linters.
+
+The linters look at your code and do two things:
+
+- ensure that your code follows the coding style adopted by the project;
+- catch a number of errors in your code.
+
+They're pretty fast, don't hesitate!
+
+```sh
+source ./env/bin/activate
 ./scripts-dev/lint.sh
 ```
 
-**Note that the script does not just test/check, but also reformats code, so you
-may wish to ensure any new code is committed first**.
+Note that the linters *will modify your files* to fix styling errors.
+Make sure that you have saved all your files.
 
-By default, this script checks all files and can take some time; if you alter
-only certain files, you might wish to specify paths as arguments to reduce the
-run-time:
+If you wish to restrict the linters to only the files changed since the last commit
+(much faster!), you can instead run:
 
+```sh
+source ./env/bin/activate
+./scripts-dev/lint.sh -d
 ```
+
+Or if you know exactly which files you wish to lint, you can instead run:
+
+```sh
+source ./env/bin/activate
 ./scripts-dev/lint.sh path/to/file1.py path/to/file2.py path/to/folder
 ```
 
-You can also provide the `-d` option, which will lint the files that have been
-changed since the last git commit. This will often be significantly faster than
-linting the whole codebase.
+## Run the unit tests.
+
+The unit tests run parts of Synapse, including your changes, to see if anything
+was broken. They are slower than the linters but will typically catch more errors.
+
+```sh
+source ./env/bin/activate
+python -m twisted.trial tests
+```
+
+If you wish to only run *some* unit tests, you may specify
+another module instead of `tests` - or a test class or a method:
+
+```sh
+source ./env/bin/activate
+python -m twisted.trial tests.rest.admin.test_room tests.handlers.test_admin.ExfiltrateData.test_invite
+```
+
+If your tests fail, you may wish to look at the logs:
+
+```sh
+less _trial_temp/test.log
+```
+
+## Run the integration tests.
+
+The integration tests run a full Synapse, including your changes, to
+see if anything was broken. They are slower than the unit tests but will
+typically catch more errors.
+
```

Comment:
some misunderstanding here. `tox` runs the same set of tests as `trial`. I'm not entirely sure I'd describe those tests as integration tests.

Reply:
So, is there any way to run integration tests?

If not, I suppose that I should entirely remove this subsection.

Reply:
Integration tests are run via sytest, we should probably point to the sytest repo that has it's own readme and not go into detail with it here.

Reply:
The previous version of the doc did exactly that: https://github.com/matrix-org/synapse/pull/9281/files#diff-eca12c0a30e25b4b46522ebf89465a03ba72a03f540796c979137931d8f92055L262-L265

Reply:
> Integration tests are run via sytest, we should probably point to the sytest repo that has it's own readme and not go into detail with it here.

I believe that it would be better if at least an example was in this document.

Now, I just need to understand that README.md :)